core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala

```diff
@@ -0,0 +1,396 @@
+        s"invoker${instanceId.toInt} is unhealthy",
+        akka.event.Logging.WarningLevel)
+      startTestAction(self)
+      publishHealthStatusAndStay(Unhealthy, nextStateData)
+
+    case _ -> Healthy =>
+      logging.info(this, s"invoker became healthy, stop health action proxy.")
+      unstashAll()
+      stopTestAction()
+
+      publishHealthStatusAndStay(Healthy, nextStateData)
+
+    case Offline -> Offline =>
+    // this is an initial transition due to startWith, do nothing
+
+    case _ -> newState =>
+      publishHealthStatusAndStay(newState, nextStateData)
+
+      unstashAll()
+
+  }
+
+  private def publishHealthStatusAndStay(state: InvokerState, stateData: InvokerHealthData) = {
+    stateData match {
+      case data: InvokerInfo =>
+        val invokerResourceMessage = InvokerResourceMessage(
+          state.asString,
+          data.memory.freeMemory,
+          data.memory.busyMemory,
+          data.memory.inProgressMemory,
+          instanceId.tags,
+          instanceId.dedicatedNamespaces)
+        InvokerHealthManager.useMemory = invokerResourceMessage.busyMemory + invokerResourceMessage.inProgressMemory
+        dataManagementService ! UpdateDataOnChange(InvokerKeys.health(instanceId), invokerResourceMessage.serialize)
+
+        stay using data.copy(currentInvokerResource = Some(invokerResourceMessage))
+
+      case data =>
+        logging.error(this, s"unexpected data is found: $data")
+
+        stay
+    }
+  }
+
+  initialize()
+
+  private def startTestAction(manager: ActorRef): Unit = {
+    val namespace = InvokerHealthManager.healthActionIdentity.namespace.name.asString
+    val docId = InvokerHealthManager.healthAction(instanceId).get.docid
+
+    WhiskAction.get(entityStore, docId).onComplete {
+      case Success(action) =>
+        val initialize = Initialize(namespace, action.toExecutableWhiskAction.get, "", 0, transid)
+        startHealthAction(initialize, manager)
+      case Failure(t) => logging.error(this, s"get health action error: ${t.getMessage}")
+    }
+  }
+
+  private def startHealthAction(initialize: Initialize, manager: ActorRef): Unit = {
+    healthActionProxy match {
+      case Some(proxy) =>
+        // make healthContainerProxy's status is Running, then healthContainerProxy can fetch the activation using ActivationServiceClient
+        proxy ! initialize
+      case None =>
+        val proxy = healthContainerProxyFactory(context, manager)
+        proxy ! initialize
+        healthActionProxy = Some(proxy)
+    }
+  }
+
+  def stopTestAction(): Unit = {
+    healthActionProxy.foreach {
+      healthActionProxy = None
+      _ ! GracefulShutdown
+    }
+  }
+
+  /**
+   * This method is to handle health message from ContainerProxy.pub
+   * It can induce status change.
+   *
+   * @param state  activation result state
+   * @param buffer RingBuffer to track status
+   * @return
+   */
+  def handleHealthMessage(state: Boolean, buffer: RingBuffer[Boolean]): State = {
```

Comment:
In the existing invoker supervision, the buffer includes user actions result for the ring buffer to check for system error. I don't know without seeing more of the new scheduler code, but we'll still have those checks right elsewhere? Also since this ring buffer seems to only include health check activations, can this be more aggressive? I would think you can just go unhealthy if a single one fails rather than three of the last ten.

Reply:
Currently, i used a simple method to check invoker's health status.
did't check for system error.

Maybe after this pr merged, we can submit another pr to optimize it. (I din't have much more good idea for it.)

Reply:
Currently, i used a simple method to check invoker's health status.
did't check for system error.

Reply:
We just preserved the existing health check logic but I think we can make it configurable.
Each container proxy is supposed to send a message indicating a failed result to InvokerHealthManager